### PR TITLE
Training refactoring

### PR DESCRIPTION
**Pull request by apbard**

### CODE REVIEW

**Comment by apbard** [comment]
logging is not working as expected

File: examples/nanoChatGPT/models/actor_critic.py [hunk 1]
```diff
@@ -15,6 +15,18 @@
...
+        super().__init__()
+        self.in_keys = []
+        self.out_keys = []
+        self.prefix = prefix
+
+    def forward(self, td):
+        print(self.prefix, td)
+        return td
+
+
```
Comment (vmoens) [inline examples/nanoChatGPT/models/actor_critic.py:29]: I designed that for debugging but if you're not using it we can drop it

File: examples/nanoChatGPT/models/actor_critic.py [hunk 2]
```diff
@@ -42,8 +54,11 @@ def __init__(self, base_model):
                 return_log_prob=True,
             ),
         )
-        value_head = TensorDictModule(
-            value_head, in_keys=["x"], out_keys=["state_value"]
+        value_head = TensorDictSequential(
+            TensorDictModule(value_head, in_keys=["x"], out_keys=["state_value"]),
+            TensorDictModule(
+                lambda x: x[:, -1, :], in_keys=["state_value"], out_keys=["state_value"]
+            ),
```
Comment (vmoens) [inline examples/nanoChatGPT/models/actor_critic.py:61]: does this indexing make sense to you?

File: torchrl/collectors/collectors.py [hunk 3]
```diff
@@ -300,7 +300,7 @@ def update_policy_weights_(
         if policy_weights is not None:
             self.policy_weights.apply(lambda x: x.data).update_(policy_weights)
         elif self.get_weights_fn is not None:
-            self.policy.load_state_dict(self.get_weights_fn())
+            self.policy_weights.apply(lambda x: x.data).update_(self.get_weights_fn())
```
Comment (vmoens) [inline torchrl/collectors/collectors.py:303]: This has been integrated into main, if you merge it you can drop that

File: torchrl/objectives/ppo.py [hunk 4]
```diff
@@ -434,11 +434,14 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
...
-                f"and {log_weight.shape})"
-            )
+        if advantage.shape != log_weight.shape:
+            try:
+                log_weight = log_weight.view_as(advantage)
+            except Exception as err:
+                raise RuntimeError(
+                    f"advantage.shape and log_weight.shape do not match (got {advantage.shape} "
+                    f"and {log_weight.shape})"
+                ) from err
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:444]: I would drop this but change the indexing in the model as I indicated in https://github.com/tcbegley/rl/pull/34 [:, [-1], :] -> [:, -1, :]
like that we would not have to change PPO

File: examples/nanoChatGPT/train_rlhf.py [hunk 5]
```diff
@@ -57,56 +96,102 @@ def forward(self, tensordict):
...
+    test_rewards = []
     losses = []
 
+    collector = SyncDataCollector(
+        env,
+        actor.eval(),
+        frames_per_batch=ep_length * num_envs,
+        total_frames=total_frames,
+        device=device,
+        storing_device="cpu",
```
Comment (vmoens) [inline examples/nanoChatGPT/train_rlhf.py:147]: we could store on GPU I think, but "cpu" may be safer

**Review by vmoens — APPROVED** [review]
Looks good!

**Review by tcbegley — APPROVED** [review]
LGTM!

**Pull request merged**